tools/ami-cleanup/src/internal/cleanup.go

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2024 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package internal
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/account"
+	accountTypes "github.com/aws/aws-sdk-go-v2/service/account/types"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/gravitational/trace"
+	"github.com/relvacode/iso8601"
+)
+
+type ApplicationInstance struct {
+	shouldDoDryRun bool
+
+	// Dependency injection for testing
+	accountClientGenerator func(cfg *aws.Config) IAccountApi
+	ec2ClientGenerator     func(cfg *aws.Config) IEc2Api
+}
+
+// Creates a new instance of the tool
+func NewApplicationInstance(doDryRun bool) *ApplicationInstance {
+	return &ApplicationInstance{
+		shouldDoDryRun:         doDryRun,
+		accountClientGenerator: NewAccountApi,
+		ec2ClientGenerator:     NewEc2Api,
+	}
+}
+
+// Performs cleanup
+func (ai *ApplicationInstance) Run(ctx context.Context) error {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to load AWS credentials")
+	}
+
+	enabledRegions, err := ai.getEnabledRegions(ctx, ai.accountClientGenerator(&cfg))
+	if err != nil {
+		return trace.Wrap(err, "failed to get enabled regions")
+	}
+
+	if len(enabledRegions) == 0 {
+		return nil
+	}
+
+	totalSpaceRecovered := int32(0)
+	totalImagesDeleted := 0
+	// This would probably run much faster with channels/concurrency, but there isn't much
+	// need with how often this is expected to be ran
+	for _, enabledRegion := range enabledRegions {
+		spaceRecovered, imagesDeleted, err := ai.cleanupRegion(ctx, cfg, *enabledRegion.RegionName)
+		if err != nil {
+			return trace.Wrap(err, "failed to clean up images in region %q", *enabledRegion.RegionName)
+		}
+
+		totalSpaceRecovered += spaceRecovered
+		totalImagesDeleted += imagesDeleted
+	}
+
+	log.Printf("Deleted %d GiB of %d images across %d regions", totalSpaceRecovered, totalImagesDeleted, len(enabledRegions))
+	return nil
+}
+
+// Cleans up the images in the given region. Returns the total amount of space recovered, number of images deleted,
+// and any error.
+func (ai ApplicationInstance) cleanupRegion(ctx context.Context, cfg aws.Config, regionName string) (int32, int, error) {
+	// A new client must be created for each region
+	cfg.Region = regionName
```

Comment:
Functions with side effects like this that modify the passed in argument are usually an anti-pattern, can we avoid doing that? E.g. just have a helper method that returns a client for a specified region.

Reply:
Passing the struct by value does create a new copy of the config. See https://go.dev/play/p/yS0-otWZ4KA as an example. The original `cfg` value that is passed upon call to `cleanupRegion` is not modified.